### ISSUE TITLE
Fail !!!  Step=24 dt=0.2501126680820332

### ISSUE THREAD

**Issue by andreasvarga** [mikelehu/IRK.jl#42]
I am occasionally encountering an error of the form

`Fail !!!  Step=24 dt=0.2501126680820332`

I am using a call as below

   `   sol = solve(prob, IRKGaussLegendre.IRKGL16(); adaptive = true, reltol = 1.e-10, abstol = 1.e-10, save_everystep = false)`

with 

  ` prob = ODEProblem(fclyap!, u0, tspan)`

and `fclyap!` is a function to efficiently evaluate the right hand side of a `nxn` matrix Lyapunov differential equation (only the upper-triangular `n(n+1)/2` elements are used). 

I guess it could be related to some convergence issues. I would appreciate any hint how to try to manage  such a case.

This is the whole error message (due to evaluating` sol(tf)`):

```
Fail !!!  Step=24 dt=0.2501126680820332
Fail
ERROR: UndefRefError: access to undefined reference
Stacktrace:
  [1] getindex
    @ .\array.jl:861 [inlined]
  [2] interpolation(tval::Float64, id::SciMLBase.LinearInterpolation{Vector{Float64}, Vector{Vector{Float64}}}, idxs::Nothing, deriv::Type{Val{0}}, p::SciMLBase.NullParameters, continuity::Symbol) 
    @ SciMLBase C:\Users\Andreas\.julia\packages\SciMLBase\xWByK\src\interpolation.jl:237
  [3] (::SciMLBase.LinearInterpolation{Vector{Float64}, Vector{Vector{Float64}}})(tvals::Float64, idxs::Nothing, deriv::Type, p::SciMLBase.NullParameters, continuity::Symbol)
    @ SciMLBase C:\Users\Andreas\.julia\packages\SciMLBase\xWByK\src\interpolation.jl:61
  [4] AbstractODESolution
    @ C:\Users\Andreas\.julia\packages\SciMLBase\xWByK\src\solutions\ode_solutions.jl:62 [inlined]
  [5] #_#342
    @ C:\Users\Andreas\.julia\packages\SciMLBase\xWByK\src\solutions\ode_solutions.jl:53 [inlined]
  [6] AbstractODESolution (repeats 2 times)
    @ C:\Users\Andreas\.julia\packages\SciMLBase\xWByK\src\solutions\ode_solutions.jl:53 [inlined]
  [7] tvclyap(A::PeriodicFunctionMatrix{:c, Float64}, C::PeriodicFunctionMatrix{:c, Float64}, tf::Float64, t0::Float64; adj::Bool, solver::String, reltol::Float64, abstol::Float64)
    @ PeriodicSystems C:\Users\Andreas\Documents\software\Julia\PeriodicSystems.jl\src\psclyap.jl:239
  [8] pgclyap(A::PeriodicFunctionMatrix{:c, Float64}, C::PeriodicFunctionMatrix{:c, Float64}, K::Int64; adj::Bool, solver::String, reltol::Float64, abstol::Float64)
    @ PeriodicSystems C:\Users\Andreas\Documents\software\Julia\PeriodicSystems.jl\src\psclyap.jl:178
  [9] psh2norm(psys::PeriodicStateSpace{PeriodicFunctionMatrix{:c, Float64}}, K::Int64; smarg::Int64, fast::Bool, offset::Float64, solver::String, reltol::Float64, abstol::Float64)
    @ PeriodicSystems C:\Users\Andreas\Documents\software\Julia\PeriodicSystems.jl\src\psanalysis.jl:313
 [10] top-level scope
    @ REPL[1]:1
```

**Comment by andreasvarga** [comment]
The same works if I am calling with

`
   sol = solve(prob, IRKGaussLegendre.IRKGL16(); adaptive = true, reltol, abstol, save_everystep = false, dt = 0.0001)`

Values of `dt = 0.1, 0.01`, and `0.001` all fail in a similar way (with different `  Step` values).

Is any way to automatically choose a feasible value for `dt`  ?

**Comment by ChrisRackauckas** [comment]
> Is any way to automatically choose a feasible value for dt ?

Not with this library.

What tolerance? Did you try decreasing it?

**Comment by andreasvarga** [comment]
This works: 
`sol = solve(prob, IRKGaussLegendre.IRKGL16(); adaptive = true, reltol = 1.e-10, abstol = 1.e-10, save_everystep = false, dt = 0.0001)`

However, it fails for  `reltol = 1.e-11, abstol = 1.e-11` and for  `reltol = 1.e-9, abstol = 1.e-9` (this last one even crashing Julia!)

**Comment by mikelehu** [comment]
It's a problem with the adaptive step size method. You can try increasing the value of the "maxtrials" parameter.
This parameter controls the maximum number of attempts to accept the adaptive step size, and the default value is 3.
e.g :
sol = solve(prob, IRKGaussLegendre.IRKGL16(maxtrials=4); adaptive = true, reltol = 1.e-10, abstol = 1.e-10, save_everystep = false)

**Comment by andreasvarga** [comment]
This doesn't work for me.

Without adaptive step size, there are no problems for dt = 0.0001..0.3.

Thanks for your time.

**Comment by andreasvarga** [comment]
Just a last comment: It would be useful for users if the code would fail in a controlled way (e.g., by issuing a meaningful error message).  In this case, it could be possible to use a `try..catch` scheme to switch to the non-adaptive version. Presently, I am encountering most of time failures which crash Julia 1.7 on Windows (without even seeing the error messages).

**Comment by mikelehu** [comment]
We appreciate your feedback and will analyze the code to resolve these
issues.
Perhaps, now it may be useful for you to know that when the code fails it
returns sol.retcode=Failure and otherwise sol.retcode=Success

Hau idatzi du Andreas Varga ***@***.***) erabiltzaileak (2022
urr. 9, ig. (11:11)):

**Comment by andreasvarga** [comment]
Thanks. I included in my code a test for `sol.retcode` and in this way I can avoid troubles. I am calling first the adaptive version (is this a good idea?), and in the case of failure, I am switching to the fixed step version. Just for your information: When I am calling the code in a loop to perform the discretization of periodic Lyapunov differential equations in `K` points, a typical run for `K = 100`, now produces (sorry for the large output):
```
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=470 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=48 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=482 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=2109 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=474 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=480 dt=NaN
Fail
sol.retcode = :Failure
Fail !!!  Step=485 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=480 dt=NaN
Fail
sol.retcode = :Failure
Fail !!!  Step=489 dt=NaN
Fail
sol.retcode = :Failure
Fail !!!  Step=486 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=478 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=482 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=480 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=506 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=482 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=474 dt=NaN
Fail
sol.retcode = :Failure
Fail !!!  Step=485 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=484 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=482 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
Fail !!!  Step=496 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
Fail !!!  Step=482 dt=NaN
Fail
sol.retcode = :Failure
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success
sol.retcode = :Success    

```

**Comment by mikelehu** [comment]
We think you're better off using the adaptive step implementation. Anyway,  we are concerned that with the v0.2.0  release, the results of the fixed steps implementation were not correct. We have fixed a bug in the fixed step length implementation: when the integration reaches its maximum number of iterations, it must stop and return sol.retcode=:Failure. 

Therefore, we recommend that you update (*) to the latest version of IRKGaussLegendre.jl v0.2.1 and repeat your experiments. We are sorry and would appreciate your feedback after the new testing.

*  pkg>up IRKGaussLegendre

M.

**Comment by andreasvarga** [comment]
Unfortunately, there are still some failures (24 out of 100 calls) as can be seen below :

```
Failure (adaptive step): maximum number of trials=5 at step=492 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=470 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=48 dt=NaN 
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=464 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=2109 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=474 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=485 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=489 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=486 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=478 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=503 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=474 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=476 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=487 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=484 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=496 dt=NaN
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
```

**Comment by mikelehu** [comment]
Can you send me the information (problem, initial value,...) of one of the cases where it fails to debug the implementation?

**Comment by andreasvarga** [comment]
I tried to make a MWE to produce some error cases. The code below performs as follows on my computer:
```
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
i = 2
Failure (adaptive step): maximum number of trials=5 at step=492 dt=NaN
i = 4
Failure (adaptive step): maximum number of trials=5 at step=479 dt=NaN
i = 9
Failure (adaptive step): maximum number of trials=5 at step=460 dt=NaN
i = 23
Failure (adaptive step): maximum number of trials=5 at step=2144 dt=NaN
i = 26
Failure (adaptive step): maximum number of trials=5 at step=479 dt=NaN
i = 52
Failure (adaptive step): maximum number of trials=5 at step=490 dt=NaN
i = 54
Failure (adaptive step): maximum number of trials=5 at step=494 dt=NaN
i = 55
Failure (adaptive step): maximum number of trials=5 at step=478 dt=NaN
i = 56
Failure (adaptive step): maximum number of trials=5 at step=478 dt=NaN
i = 57
Failure (adaptive step): maximum number of trials=5 at step=488 dt=NaN
i = 58
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
i = 59
Failure (adaptive step): maximum number of trials=5 at step=496 dt=NaN
i = 60
Failure (adaptive step): maximum number of trials=5 at step=484 dt=NaN
i = 61
Failure (adaptive step): maximum number of trials=5 at step=477 dt=NaN
i = 66
Failure (adaptive step): maximum number of trials=5 at step=491 dt=NaN
i = 67
Failure (adaptive step): maximum number of trials=5 at step=482 dt=NaN
i = 70
Failure (adaptive step): maximum number of trials=5 at step=476 dt=NaN
i = 71
Failure (adaptive step): maximum number of trials=5 at step=484 dt=NaN
i = 75
Failure (adaptive step): maximum number of trials=5 at step=484 dt=NaN
i = 78
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
i = 79
Failure (adaptive step): maximum number of trials=5 at step=491 dt=NaN
i = 89
Failure (adaptive step): maximum number of trials=5 at step=480 dt=NaN
i = 91
Failure (adaptive step): maximum number of trials=5 at step=479 dt=NaN
i = 92
Failure (adaptive step): maximum number of trials=5 at step=483 dt=NaN
i = 94
Failure (adaptive step): maximum number of trials=5 at step=474 dt=NaN
i = 95
Failure (adaptive step): maximum number of trials=5 at step=487 dt=NaN
i = 97

```

```
# test example 
using IRKGaussLegendre
function muladdcsymt!(y::AbstractVector, x::AbstractVector, A::AbstractMatrix, C::AbstractMatrix)
    n = size(A, 1)
    X = reshape(x,n,n)
    # A*X + X*A' + C
    y[:] .= (A*X+X*A'+C)[:]
    return nothing
end
A = t -> [-1-sin(2*t)^2 2-0.5*sin(4*t); -2-0.5*sin(4*t) -1-cos(2*t)^2]
γ = t -> mod(t,float(pi)) < pi/2 ? sin(2*t) : 0 
B = t -> [0; 1-2*0.5*γ(t)]
K = 100
Ts = pi/K
dt = Ts/100
ftclyap!(du,u,p,t) = muladdcsymt!(du, u, A(t), B(t)*B(t)')
i1 = 1; i2 = K  # for all cases
#i1 = 2; i2 = 2 # for an error case on my machine
for i = i1:i2
    u0 = zeros(2*2) 
    tspan = ((i-1)*Ts,i*Ts)
    prob = ODEProblem(ftclyap!, u0, tspan)
    sol = solve(prob, IRKGaussLegendre.IRKGL16(); adaptive = true, reltol=1.e-10, abstol=1.e-10, save_everystep = false)
    # @show sol.retcode
    if sol.retcode == :Failure
       @show i
       sol = solve(prob, IRKGaussLegendre.IRKGL16(); adaptive = false, reltol=1.e-10, abstol=1.e-10, save_everystep = false, dt)
    end
    #@show sol[end] # solution must be symmetric
end

```

**Comment by mikelehu** [comment]
We have found that our adaptive stepping implementation sometimes does not correctly select the initial step length. Now, we are currently working on a new strategy for adaptive steps that resolves these issues and we hope to add it soon.

Meanwhile, by specifying the following options in the adaptive step implementation: dt as the initial step length and abstol=1.e-30, all executions complete successfully (see code below).

`     
       # test example 

     function muladdcsymt!(y::AbstractVector, x::AbstractVector, A::AbstractMatrix, C::AbstractMatrix)

       n = size(A, 1)

       X = reshape(x,n,n)

       # A*X + X*A' + C

       y[:] .= (A*X+X*A'+C)[:]

      return nothing

    end

    A= t -> [-1-sin(2*t)^2 2-0.5*sin(4*t); -2-0.5*sin(4*t) -1-cos(2*t)^2]

    γ = t -> mod(t,float(pi)) < pi/2 ? sin(2*t) : 0 

    B = t -> [0; 1-2*1/2*γ(t)]

    K = 100

    Ts = pi/K

    dt = Ts/100

    ftclyap!(du,u,p,t) = muladdcsymt!(du, u, A(t), B(t)*B(t)')

    i1 = 1; i2 = K  # for all cases

    #i1 = 2; i2 = 2 # for an error case on my machine

    for i = i1:i2

        u0 = zeros(2*2) 

        tspan = ((i-1)*Ts,i*Ts)

        prob = ODEProblem(ftclyap!, u0, tspan)

        sol = solve(prob, IRKGaussLegendre.IRKGL16(); dt, adaptive = true, reltol=1.e-10, abstol=1.e-30, save_everystep = false)

        if sol.retcode == :Failure  @show i end

    end`

**Comment by andreasvarga** [comment]
Yes, indeed it works. What is the role of `dt` in the case of the adaptive strategy?

**Comment by mikelehu** [comment]
Actually, the automatic choice of the initial step size is independent of the adaptive strategy. I meant that we are going to improve both strategies in the next release of IRKGL16

**Comment by andreasvarga** [comment]
Are any news about the progress of your work? 

Recently, when I am using the symplectic solver I also receive the following message:
`
WARNING: DiffEqBase.DEStats is deprecated, use Stats instead. `

**Comment by ChrisRackauckas** [comment]
on v0.2.2?

**Comment by andreasvarga** [comment]
Yes. See

https://github.com/andreasvarga/PeriodicSystems.jl/actions/runs/4531394137/jobs/7981484361

**Comment by andreasvarga** [comment]
Yes. See

https://github.com/andreasvarga/PeriodicSystems.jl/actions/runs/4531394137/jobs/7981484361

**Comment by ChrisRackauckas** [comment]
Fixed the deprecation on v0.2.3

**Comment by mikelehu** [comment]
Andrea, thanks for your interest.
We've been working hard on the new adaptive steps strategy for the past few months, but we've run into some unexpected difficulties and we have decided to reject this idea. In the next new version, we will add an IRKGL-simd (vectorized) implementation to the IRKGaussLegendre.jl package for double-precision calculations, which is transparent to the user and will significantly improve the efficiency of the method.

And thanks Chris, for fixing the problem so quickly.
